tony-core/src/main/java/com/linkedin/tony/ReusablePort.java

```diff
@@ -0,0 +1,160 @@
+   * SO_REUSEPORT.
+   * @return the created port
+   */
+  static ReusablePort create() throws IOException, InterruptedException {
+    ReusablePort reusablePort = create(getAvailablePort());
+    return reusablePort;
+  }
+
+  /**
+   * Creates a binding port with netty library which has built-in port reuse support.
+   * <p>port reuse feature is detailed in:
+   * <a href="https://lwn.net/Articles/542629/">https://lwn.net/Articles/542629/</a>
+   * </p>
+   *
+   * @param port the port to bind to, cannot be 0 to pick a random port. Since another tony
+   *             executor can bind to the same port when port 0 and SO_REUSEPORT are used together.
+   * @return the binding port
+   * @throws BindException if fails to bind to any port
+   * @throws InterruptedException if the thread waiting for incoming connection is interrupted
+   */
+  @VisibleForTesting
+  static ReusablePort create(int port) throws InterruptedException,
+      BindException {
+    // Why creating connection with port reuse using netty instead of native socket library
+    //(https://docs.oracle.com/javase/8/docs/api/java/net/Socket.html)?
+    // - Tony's default Java version is 8 and port reuse feature is only available in Java 9+:
+    //   https://docs.oracle.com/javase/9/docs/api/java/net/StandardSocketOptions.html#SO_REUSEPORT.
+    //
+    // Why not upgrading Tony to Java 9+ given port reuse is supported in Java 9+?
+    // - In Linkedin, as of now(2020/08), only Java 8 and 11 are officially supported, but Java 11
+    //   introduces incompatibility with Play version tony-portal
+    //   (https://github.com/linkedin/TonY/tree/master/tony-portal) is using. Upgrading Play to a
+    //   Java 11-compatible version requires non-trivial amount of effort.
+
+    Preconditions.checkArgument(port > 0, "port must > 0");
+    final EventLoopGroup bossGroup = new EpollEventLoopGroup();
+    ServerBootstrap b = new ServerBootstrap();
+    ChannelFuture future = null;
+
+    try {
+      b.group(bossGroup)
+          .channel(EpollServerSocketChannel.class)
+          .childHandler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            public void initChannel(SocketChannel ch) {
+            }
+          }).option(EpollChannelOption.SO_REUSEPORT, true)
+          .option(ChannelOption.SO_KEEPALIVE, true);
+
+      // Why not using port 0 here which lets kernel pick an available port?
+      // - Since another tony executor can bind to the same port when port 0 and SO_REUSEPORT are
+      //   used together. See how a port is selected by kernel based on a free-list and socket
+      //   options: https://idea.popcount.org/2014-04-03-bind-before-connect/#port-allocation.
+
+      // Note it's still slightly possible that another tony processes grabs the same port after
+      // {@link #getAvailablePort()}, leading to two tensorflow process uses the same port,
+      // though it would be rare.
+      future = b.bind(port).await();
```

Comment:
Can we do a retry on another port in the rare case that the port is already bound?

Reply:
good idea. updated.